Added StateHandler

ivargruneau opened the pull request:
Added `StateHandler` initi

functions

- `initStateHandler` takes a string as a parameter, sets up the Statehandler and sets the `TagId`
- `lock` locks the mutex
- `unLock` unlocks the mutex
- `getReading` takes a string as a parameter and returns a reading with a matching `TagId` primarily implemented for testing
- `getState` returns `State` with all the readings contained in the internal map of the Statehandler.

Changes
- `insert.go` 
- - `InsertMultipleReadings` now only takes a `State` as a parameter. Usage example: `stateHandler.InsertMultipleReadings(state)`
- - `InsertSingleReading` now only takes a `Reading` as a parameter. Now uses the `lock` and `unLock` method from the Statehandler. Usage example `stateHandler.InsertSingleReading(reading)`

- `insert_test.go` updated the test cases to work with the Statehandler

- `Tag.proto` changed `State.Readings` from a map to `repeated` Readings`

asta987 reviewed (changes requested):
In **insert_test.go**, the test **TestInsertMultipleReading** fails when I run **go test** in the terminal.

arianfiftyone commented on pkg/protocol/Tag.proto:
```diff
@@ -9,8 +9,8 @@ import "google/protobuf/timestamp.proto";
-    map<string, Reading> readings   = 2; // List of readings made by reporter
-}
+    repeated Reading    = 2; // List of readings made by reporter
+}readings
```
Change to `repeated Reading readings` inside of message definition.

arianfiftyone commented on pkg/protocol/utlis/insert_test.go:
Test passes the majority of times, but fails since expected and actual reading are not the same. Might be related to the first commend in the message definition `message State`.

arianfiftyone reviewed (changes requested):
Ty the requested changes and run the test again to make sure it always passes. Otherwise it looks good!

ivargruneau commented:
Changed state readings to: `repeated Reading readings`.
Fixed the issue of tests sometimes failing, the issue stemmed from the `getState()` function in `Statehandler`. Since iteration order over a map in Go varies the returned states readings order would be random. Fixed the problem by sorting the list before comparing them in the tests.
